### PR TITLE
Option for disabling hashing

### PR DESCRIPTION
**Pull request by chromy96**
Pull request contains option to define property in config file (using typesafe config) that will skip hashing of files. Configuration is false by default (defined in reference.conf).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -70,6 +70,20 @@ public class DirectoryWatchingUtility {
 }
 ```
 
+## Configuration
+
+By default, DirectoryWatcher will try to prevent duplicate events (e.g. Windows will emit duplicate modify events when file is changed). This is done by creating hash for every file encountered and keeping that hash in memory.
```
Comment (gmethvin) [inline README.md]: The text looks good but you are missing articles in a few places:

Comment (gmethvin) [inline README.md]: The text looks good but you are missing articles in a few places. It should be:

"By default, DirectoryWatcher will try to prevent duplicate events (e.g. Windows will emit duplicate modify events when **a** file is changed). This is done by creating **a** hash for every file encountered and keeping that hash in memory."

File: README.md [hunk 2]
```diff
@@ -70,6 +70,20 @@ public class DirectoryWatchingUtility {
...
+This might result in slower performance, because library has to calculate the hash of the entire file. In addition, some events are not emited, because file was locked by another process when library tried to calculate the hash.
```
Comment (gmethvin) [inline README.md]: Same thing: "*the* library", "*the* file".

I think it would be better to state the second sentence as: "In addition, some events may not be emitted if DirectoryWatcher encounters a file that is locked by another process while computing the hash."

File: README.md [hunk 3]
```diff
@@ -70,6 +70,20 @@ public class DirectoryWatchingUtility {
 }
 ```
 
+## Configuration
+
+By default, DirectoryWatcher will try to prevent duplicate events (e.g. Windows will emit duplicate modify events when file is changed). This is done by creating hash for every file encountered and keeping that hash in memory.
+This might result in slower performance, because library has to calculate the hash of the entire file. In addition, some events are not emited, because file was locked by another process when library tried to calculate the hash. 
+
+To avoid performance hit (and possible event loss), you can add `application.conf` file in your resource folder and set property `io.methvin.prevent-file-hashing` to `true`.
```
Comment (gmethvin) [inline README.md]: This is sort of redundant with what you said above. I'd just say "To disable hashing of files, you can ..."

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 4]
```diff
@@ -245,16 +254,21 @@ private void register(Path directory, boolean useFileTreeModifier) throws IOExce
...
 
   private void notifyCreateEvent(Path path, int count) throws IOException {
-    HashCode newHash = PathUtils.hash(path);
-    if (newHash == null) {
-      logger.debug("Failed to hash created file [{}]. It may have been deleted.", path);
-      return;
-    }
-    // Notify for the file create if not already notified
-    if (!pathHashes.containsKey(path)) {
+    if(settings.isPreventFileHashing() && Files.isRegularFile(path)) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: Why are you using `isDirectory` in the other place and `isRegularFile` here?

Also let's keep the formatting consistent and use `if (` rather than `if(`

**Comment by gmethvin** [comment]
Yeah, if you have other ideas for the configuration I'm open to hearing them. A builder would be not so bad:

```java
DirectoryWatcher.builder()
  .enableHashing(false)
  .paths(paths)
  .listener(evt => {
    System.out.println("Got event: " + evt);
  })
  .build()
  .watch();
```
I think for now we can just add the additional parameter.

**Comment by chromy96** [comment]
I've pushed the change that addresses your comments.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 5]
```diff
@@ -59,15 +59,24 @@
 
   // this is set to true/false depending on whether recursive watching is supported natively
   private Boolean fileTreeSupported = null;
+  private Boolean enableFileHashing = true;
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: This doesn't need to be a `Boolean`. It can just be a regular `boolean`. The other one is `Boolean` since I needed to handle the uninitialized state.

File: README.md [hunk 6]
```diff
@@ -70,6 +70,16 @@ public class DirectoryWatchingUtility {
 }
 ```
 
+## Configuration
+
+By default, DirectoryWatcher will try to prevent duplicate events (e.g. Windows will emit duplicate modify events when a file is changed). This is done by creating a hash for every file encountered and keeping that hash in memory.
+This might result in slower performance, because the library has to calculate the hash of the entire file. In addition, some events may not be emitted if DirectoryWatcher encounters a file that is locked by another process while computing the hash.
+
+To disable hashing, you can set `enableHashing` parameter to `false` when calling `DirectoryWatcher.create`
```
Comment (gmethvin) [inline README.md]: ...the `enableFileHashing` parameter...

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 7]
```diff
@@ -59,15 +59,24 @@
...
+  }
+
+  public static DirectoryWatcher create(List<Path> paths, DirectoryChangeListener listener, Boolean enableFileHashing) throws IOException {
     boolean isMac = System.getProperty("os.name").toLowerCase().contains("mac");
     WatchService ws = isMac ? new MacOSXListeningWatchService() : FileSystems.getDefault().newWatchService();
-    return new DirectoryWatcher(paths, listener, ws);
+    return new DirectoryWatcher(paths, listener, ws, enableFileHashing);
   }
 
   public DirectoryWatcher(List<Path> paths, DirectoryChangeListener listener, WatchService watchService) throws IOException {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: This can delegate to the other constructor.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 8]
```diff
@@ -59,15 +59,24 @@
...
   // this is set to true/false depending on whether recursive watching is supported natively
   private Boolean fileTreeSupported = null;
+  private Boolean enableFileHashing = true;
 
   public static DirectoryWatcher create(Path path, DirectoryChangeListener listener) throws IOException {
-    return create(Collections.singletonList(path), listener);
+    return create(Collections.singletonList(path), listener, Boolean.TRUE);
+  }
+
+  public static DirectoryWatcher create(Path path, DirectoryChangeListener listener, Boolean enableFileHashing) throws IOException {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: I'd just omit this factory method.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 7 above)
Comment (gmethvin) [reply core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: Or maybe just omit it entirely, since the recommended way to create a `DirectoryWatcher` is through the `create` call. I'm not too worried about breaking API changes since the surface area is very small.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 9]
```diff
@@ -245,16 +275,21 @@ private void register(Path directory, boolean useFileTreeModifier) throws IOExce
...
 
   private void notifyCreateEvent(Path path, int count) throws IOException {
-    HashCode newHash = PathUtils.hash(path);
-    if (newHash == null) {
-      logger.debug("Failed to hash created file [{}]. It may have been deleted.", path);
-      return;
-    }
-    // Notify for the file create if not already notified
-    if (!pathHashes.containsKey(path)) {
+    if(enableFileHashing || Files.isDirectory(path)){
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: `) {`

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 7 above)
Comment (chromy96) [reply core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: I thought of removing it, but didn't want to brake the API. I will remove this constructor then.

**Comment by jvican** [comment]
I think that this option should be `false` by default. There are many applications that have idempotent functions (or functions checking these hashes and de-duplicating the evaluation). In the spirit of keeping backwards compatibility, I think it's important that we try to avoid these "insidious" performance bugs that can occur by accident.

**Comment by chromy96** [comment]
@gmethvin Last 2 commits address the your comments

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 10]
```diff
@@ -59,28 +59,35 @@
 
   // this is set to true/false depending on whether recursive watching is supported natively
   private Boolean fileTreeSupported = null;
+  private boolean enableFileHashing = true;
 
   public static DirectoryWatcher create(Path path, DirectoryChangeListener listener) throws IOException {
-    return create(Collections.singletonList(path), listener);
+    return create(Collections.singletonList(path), listener, Boolean.TRUE);
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: Basically everywhere you use `Boolean` instances here you can use `boolean`, so this can be `true`.

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherTest.java [hunk 11]
```diff
@@ -148,8 +148,7 @@ protected void runWatcher(Path directory, WatchService watchService) throws Exce
     List<FileSystemAction> actions = fileSystem.actions();
 
     TestDirectoryChangeListener listener = new TestDirectoryChangeListener(directory, actions);
-    DirectoryWatcher watcher = new DirectoryWatcher(Collections.singletonList(directory), listener, watchService);
-
+    DirectoryWatcher watcher = new DirectoryWatcher(Collections.singletonList(directory), listener, watchService, Boolean.TRUE);
```
Comment (gmethvin) [inline core/src/test/java/io/methvin/watchservice/DirectoryWatcherTest.java]: `true`. Also did you try running the tests with a value of `false` here? It would be interesting to see which tests end up with duplicate events.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 12]
```diff
@@ -59,28 +59,35 @@
...
   }
 
   public static DirectoryWatcher create(List<Path> paths, DirectoryChangeListener listener) throws IOException {
+    return create(paths, listener, Boolean.TRUE);
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: `true`

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 13]
```diff
@@ -245,16 +256,21 @@ private void register(Path directory, boolean useFileTreeModifier) throws IOExce
...
 
   private void notifyCreateEvent(Path path, int count) throws IOException {
-    HashCode newHash = PathUtils.hash(path);
-    if (newHash == null) {
-      logger.debug("Failed to hash created file [{}]. It may have been deleted.", path);
-      return;
-    }
-    // Notify for the file create if not already notified
-    if (!pathHashes.containsKey(path)) {
+    if(enableFileHashing || Files.isDirectory(path)) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: `if (`

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java [hunk 14]
```diff
@@ -30,7 +35,7 @@
   public void setUp() throws IOException {
     this.tmpDir = Files.createTempDirectory(null);
     this.recorder = new EventRecorder();
-    this.watcher = DirectoryWatcher.create(this.tmpDir, this.recorder);
+    this.watcher = DirectoryWatcher.create(Collections.singletonList(this.tmpDir), this.recorder, false);
```
Comment (gmethvin) [inline core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java]: It might be worth running all these tests with both `true` and `false` to make sure both work. What do you think?

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java [hunk 15]
```diff
@@ -108,7 +113,37 @@ public void moveSubDirectory() throws IOException, ExecutionException, Interrupt
...
+        // Expected exception.
+      }
+
+      assertTrue(
+          "Create event for the child file was notified",
+          this.recorder.events.stream().anyMatch(
+              e -> e.eventType() == DirectoryChangeEvent.EventType.CREATE && e.path().getFileName().equals(child.getFileName())));
 
+    } finally {
+      if(lock != null && channel != null && channel.isOpen()){
```
Comment (gmethvin) [inline core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java]: `if (`

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 16]
```diff
@@ -161,19 +168,23 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
             }
             notifyCreateEvent(childPath, count);
           } else if (kind == ENTRY_MODIFY) {
-            // Note that existingHash may be null due to the file being created before we start listening
-            // It's important we don't discard the event in this case
-            HashCode existingHash = pathHashes.get(childPath);
+            if(enableFileHashing || Files.isDirectory(childPath)) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: `if (`

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java [hunk 17]
```diff
@@ -108,7 +113,37 @@ public void moveSubDirectory() throws IOException, ExecutionException, Interrupt
...
+    try {
+
+      File file = child.toFile();
+      channel = new RandomAccessFile(file, "rw").getChannel();
+      lock = channel.lock();
+
+      try {
+        future.get(5, TimeUnit.SECONDS);
+      } catch (TimeoutException e) {
+        // Expected exception.
```
Comment (gmethvin) [inline core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java]: You might just want to throw this since the test would probably fail otherwise, right?

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 18]
```diff
@@ -59,28 +59,35 @@
...
+
+  public static DirectoryWatcher create(List<Path> paths, DirectoryChangeListener listener, Boolean enableFileHashing) throws IOException {
     boolean isMac = System.getProperty("os.name").toLowerCase().contains("mac");
     WatchService ws = isMac ? new MacOSXListeningWatchService() : FileSystems.getDefault().newWatchService();
-    return new DirectoryWatcher(paths, listener, ws);
+    return new DirectoryWatcher(paths, listener, ws, enableFileHashing);
   }
 
-  public DirectoryWatcher(List<Path> paths, DirectoryChangeListener listener, WatchService watchService) throws IOException {
+  public DirectoryWatcher(List<Path> paths, DirectoryChangeListener listener, WatchService watchService, Boolean enableFileHashing) throws IOException {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: `boolean`

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 19]
```diff
@@ -59,28 +59,35 @@
...
   public static DirectoryWatcher create(Path path, DirectoryChangeListener listener) throws IOException {
-    return create(Collections.singletonList(path), listener);
+    return create(Collections.singletonList(path), listener, Boolean.TRUE);
   }
 
   public static DirectoryWatcher create(List<Path> paths, DirectoryChangeListener listener) throws IOException {
+    return create(paths, listener, Boolean.TRUE);
+  }
+
+  public static DirectoryWatcher create(List<Path> paths, DirectoryChangeListener listener, Boolean enableFileHashing) throws IOException {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: `boolean`

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 20]
```diff
@@ -59,28 +59,35 @@
...
     this.watchService = watchService;
     this.isMac = watchService instanceof MacOSXListeningWatchService;
     this.pathHashes = PathUtils.createHashCodeMap(paths);
     this.keyRoots = PathUtils.createKeyRootsMap();
+    this.enableFileHashing = enableFileHashing;
 
     for (Path path : paths) {
       registerAll(path);
     }
+
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: delete

**Comment by gmethvin** [comment]
@chromy96 This looks good besides the few things I mentioned in my last review. It'd be good to get this merged soon, so I'm happy to take over if you don't have the time. Thanks again for your help here.

**Comment by chromy96** [comment]
Thanks, @gmethvin
I can address the comments for the weekend, if that's not too late.

**Comment by gmethvin** [comment]
@chromy96 Yeah that's fine. I just didn't see any activity in a few weeks so I was just checking :)

**Comment by chromy96** [comment]
Great. Yeah, was little busy finishing things before vacation :)

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherTest.java (hunk 11 above)
Comment (chromy96) [reply core/src/test/java/io/methvin/watchservice/DirectoryWatcherTest.java]: I think this depends largely on the OS. In case of Win10, it's usually update events that show up multiple times. E.g. in this test, for file one.txt instead of only create and delete, we have: create, update and delete events.

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java (hunk 17 above)
Comment (chromy96) [reply core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java]: It's the same as with other tests, it is expected to timeout since the watcher is not closed.

File: core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java [hunk 21]
```diff
@@ -108,7 +133,73 @@ public void moveSubDirectory() throws IOException, ExecutionException, Interrupt
...
+      }
+
+      assertEquals("No event for the file creation expected", 0, this.recorder.events.size());
```
[the remark this answers is not in the document]
Comment (chromy96) [reply core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java]: Yes, file locking is platform-dependent and it might be difficult to reproduce in all platforms. We can enable it only in windows, where we see it's working. That way we'll have  a demonstration of a file hashing problem that this PR is trying to resolve. What do you think?

**Comment by chromy96** [comment]
@gmethvin could you retrigger the AppVeyor build? It failed due to remote service unavailability.

**Comment by gmethvin** [comment]
Released as v0.6.0.

**Comment by chromy96** [comment]
Excellent, thanks :)
